src/util/prometheus/impl/HistogramImpl.h

```diff
@@ -0,0 +1,130 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2023, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include <util/Concepts.h>
+#include <util/prometheus/OStream.h>
+
+#include <mutex>
+
+namespace util::prometheus::detail {
+
+template <typename T>
+concept SomeHistogramImpl = requires(T t) {
+    typename std::remove_cvref_t<T>::ValueType;
+    SomeNumberType<typename std::remove_cvref_t<T>::ValueType>;
+    {
+        t.observe(typename std::remove_cvref_t<T>::ValueType{1})
+    } -> std::same_as<void>;
+    {
+        t.setBuckets(std::vector<typename std::remove_cvref_t<T>::ValueType>{})
+    } -> std::same_as<void>;
+    {
+        t.serializeValue(std::string{}, std::string{}, std::declval<OStream&>())
+    } -> std::same_as<void>;
+};
+
+template <SomeNumberType NumberType>
+class HistogramImpl {
+public:
+    using ValueType = NumberType;
+
+    HistogramImpl() = default;
+    HistogramImpl(HistogramImpl const&) = delete;
+    HistogramImpl(HistogramImpl&&) = default;
+    HistogramImpl&
+    operator=(HistogramImpl const&) = delete;
+    HistogramImpl&
+    operator=(HistogramImpl&&) = default;
+
+    void
+    setBuckets(std::vector<ValueType> const& bounds)
```

Comment:
Curious , if the buckets can only be set once, how about set it in constructor? like AnyHistogram

Reply:
`HistogramImpl` have to be default constructable to be able to make `impl = detail::HistogramImpl{}` in AnyHistogram's constructor.